tests/unit/set-test.js

```diff
@@ -0,0 +1,25 @@
+var simple = require('simple-mock')
+var test = require('tape')
+
+var set = require('../../lib/set')
+var internals = set.internals
+
+test('set', function (t) {
+  t.plan(1)
+
+  var state = {}
+  var username = 'admin'
+  var password = 'secret'
+
+  simple.mock(internals, 'hashPassword').callbackWith(new Error('hashPassword error'))
+
+  set(state, username, password)
+
+  .then(function () {
+    t.error('Should reject')
+  })
+
+  .catch(function (error) {
+    t.is(error.message, 'hashPassword error', 'rejects with error from hashPassword')
+  })
+})
```

Comment:
when we do unit tests, then we try to only test one method at a time, in this case the `set` method. What you will see a lot in Hoodie’s code is a first argument `state`, which is passed in when initialising the API in the constructor, in this case here: https://github.com/hoodiehq/pouchdb-admins/blob/master/lib/api.js#L16

`set.bind(null, state)` will implicitly pass in the `state` argument each time someone calls `admins.set(username, password)`. So `state` is not accessible from the outside. It’s a style of functional programming. The benefit is that for testing, we can put in what ever `state` we want for the respective test case, which is very helpful

Check out unit tests of [hoodie-account-client](https://github.com/hoodiehq/hoodie-account-client/tree/master/test/unit) to see how we do unit tests.